src/ILLink.RoslynAnalyzer/RequiresAnalyzerBase.cs

```diff
@@ -0,0 +1,193 @@
+
+					if (ReportDangerousPatternDiagnostic (operationContext, dangerousPatterns, member))
+						return;
+
+					if (!member.HasAttribute (RequiresAttributeName))
+						return;
+
+					// Warn on the most derived base method taking into account covariant returns
+					while (member is IMethodSymbol method && method.OverriddenMethod != null && SymbolEqualityComparer.Default.Equals (method.ReturnType, method.OverriddenMethod.ReturnType))
+						member = method.OverriddenMethod;
+
+					if (TryGetRequiresAttribute (member, out AttributeData? requiresAttribute)) {
+						ReportRequiresDiagnostic (operationContext, member, requiresAttribute);
+					}
+				}
+			});
+		}
+
+		[Flags]
+		protected enum DiagnosticTargets
+		{
+			MethodOrConstructor = 0x0001,
+			Property = 0x0002,
+			Field = 0x0004,
+			Event = 0x0008,
+			All = MethodOrConstructor | Property | Field | Event
+		}
+
+		private static ISymbol FindContainingSymbol (OperationAnalysisContext operationContext, DiagnosticTargets targets)
+		{
+			var parent = operationContext.Operation.Parent;
+			while (parent is not null) {
+				switch (parent) {
+				case IAnonymousFunctionOperation lambda:
+					return lambda.Symbol;
+				case ILocalFunctionOperation local when targets.HasFlag (DiagnosticTargets.MethodOrConstructor):
+					return local.Symbol;
+				case IMethodBodyBaseOperation when targets.HasFlag (DiagnosticTargets.MethodOrConstructor):
+				case IPropertyReferenceOperation when targets.HasFlag (DiagnosticTargets.Property):
+				case IFieldReferenceOperation when targets.HasFlag (DiagnosticTargets.Field):
+				case IEventReferenceOperation when targets.HasFlag (DiagnosticTargets.Event):
+					return operationContext.ContainingSymbol;
+				default:
+					parent = parent.Parent;
+					break;
+				}
+			}
+			return operationContext.ContainingSymbol;
+		}
+
+		private void ReportRequiresDiagnostic (OperationAnalysisContext operationContext, ISymbol member, AttributeData? requiresAttribute)
+		{
+			var message = GetMessageFromAttribute (requiresAttribute);
+			var url = GetUrlFromAttribute (requiresAttribute);
+			operationContext.ReportDiagnostic (Diagnostic.Create (
+				RequiresDiagnosticRule,
+				operationContext.Operation.Syntax.GetLocation (),
+				member.ToString (),
+				message,
+				url));
+		}
+
+		protected abstract string GetMessageFromAttribute (AttributeData? requiresAssemblyFilesAttribute);
+
+		private string GetUrlFromAttribute (AttributeData? requiresAssemblyFilesAttribute)
+		{
+			var url = requiresAssemblyFilesAttribute?.NamedArguments.FirstOrDefault (na => na.Key == "Url").Value.Value?.ToString ();
+			return string.IsNullOrEmpty (url) ? "" : " " + url;
+		}
+
+		protected abstract bool TryGetRequiresAttribute (ISymbol member, out AttributeData? requiresAttribute);
+
+		protected abstract bool ReportDangerousPatternDiagnostic (OperationAnalysisContext operationContext, ImmutableArray<ISymbol> dangerousPatterns, ISymbol member);
```

Comment:
Some of these probably deserve doc comments. It's not clear to me what each method is supposed to do.